### ISSUE TITLE
How to handle publishTimeoutException()

### ISSUE THREAD

**Issue by flycast** [aws/aws-iot-device-sdk-python#84]
I have a python script that is simulating data being published in the kinds of timing and pauses that I expect to be getting in real life. It publishes messages randomly between 1 - 6 seconds. After each loop it randomly determines to take a nap of up to 600 seconds.

I am getting an expection:

> publishTimeoutException()

That exception always seems to come while it is publishing messages on the 1-6 second interval. It is publishing most messages (>99%).

What causes this message? I assume a lack of connection?
What is the best practice for handling this exception...reconnect and republish?

Here is my code with a try, exception added:

```
myMQTTClient.configureOfflinePublishQueueing(-1)  # Infinite offline Publish queueing
myMQTTClient.configureDrainingFrequency(2)  # Draining: 2 Hz
myMQTTClient.configureConnectDisconnectTimeout(60)  # 10 sec
myMQTTClient.configureMQTTOperationTimeout(20)  # 5 sec

#Variables
cycleBase = 1
setUpGap = 25

myMQTTClient.connect()
while True:
    if random.uniform(0,100) < 100-(100/setUpGap):
        #Do a cycle
        message = "{ \"timeStamp\":" + str(time.time()) + ", \"thing\":\"testing/IoT\" }"
        try:
            myMQTTClient.publish("testing/IoT", message, 1)
        except:
            print(err)
            myMQTTClient.connect()
            myMQTTClient.publish("testing/IoT", message, 1)

        
        
        time.sleep(cycleBase + random.uniform(0,5))
        print(message)
    else:
        #Do a set up
        setUpGap = random.uniform(120, 600)
        print("Set up for " + str(setUpGap))
        time.sleep(setUpGap)
    
myMQTTClient.disconnect()
```

**Comment by hedeshianaws** [comment]
@flycast I apologize for the delay in getting back to you. Which SDK version are you using?

One thing that may help in diagnosing would be to print the return codes from myMQTTClient.publish().

Also, are you hitting this block at all?
```
        except:
            print(err)
            myMQTTClient.connect()
            myMQTTClient.publish("testing/IoT", message, 1)
```

**Comment by hedeshianaws** [comment]
Thanks again for your interest in AWS IoT. I'm closing this issue due to inactivity. Please feel free to re-open if you have additional questions, or create a new issue if you encounter other problems.

**Comment by esimkowitz** [comment]
I'm having a similar issue. I'm trying to send a ~1MB payload and my publish keeps timing out. I've tried it up to 60s timeout. I'm on version 1.2.0 of the SDK and running Python 3.6.2. My error code is as follows:
```text
Publish timed out
Traceback (most recent call last):
  File ".\index.py", line 50, in <module>
    myMQTTClient.publish("test/test1", json.dumps(new_data), 1)
  File "C:\Users\XXXX\AppData\Local\Programs\Python\Python36-32\lib\site-packages\AWSIoTPythonSDK\MQTTLib.py", line 523, in publish
    return self._mqtt_core.publish(topic, payload, QoS, False)  # Disable retain for publish by now
  File "C:\Users\XXXX\AppData\Local\Programs\Python\Python36-32\lib\site-packages\AWSIoTPythonSDK\core\protocol\mqtt_core.py", line 221, in publish
    raise publishTimeoutException()
AWSIoTPythonSDK.exception.AWSIoTExceptions.publishTimeoutException
```
Any help would be greatly appreciated.

**Comment by liuszeng** [comment]
Hi @esimkowitz ,

Thank you very much for using AWS IoT Python SDK.

AWS IoT has limitations on the PUBLISH message size of 128 KB. Any PUBLISH message larger than this size will be rejected. More details can be found here (search for "Message size" on the page):
http://docs.aws.amazon.com/general/latest/gr/aws_service_limits.html#limits_iot

I see you are using synchronous publish API. Chances are that your publish request is out. However, the server disconnects you already when your outbound publish request exceeds the maximum allowed message size. Therefore, the sync call waits on a PUBACK that will never come back and eventually times out.

If you have further questions, I strongly recommend you to open up a new thread in this repository so that it can be easier to be picked up and tracked internally for us.

Thanks,
Liusu

**Comment by tanonymous22** [comment]
I am also having a similar issue. I have created a button and on clicking that my func gets called and it publishes a message. On the other hand, my subscribe keeps running for infinite time. Mine subscribe script runs but my publish script does not. after sometime it shows connection timeout message. And it does not even prints the message.

publish.py
from AWSIoTPythonSDK.MQTTLib import AWSIoTMQTTClient
from tkinter import *
import time
myMQTTClient = AWSIoTMQTTClient('Client1')
myMQTTClient.configureEndpoint('host',8883)
myMQTTClient.configureCredentials(rootCAPath, privateKeyPath, certificatePath)


master = Tk()
global count
def publish():
    myMQTTClient.configureAutoReconnectBackoffTime(1, 32, 20)
    myMQTTClient.configureOfflinePublishQueueing(-1)
    myMQTTClient.configureDrainingFrequency(2)
    myMQTTClient.configureConnectDisconnectTimeout(10)
    myMQTTClient.configureMQTTOperationTimeout(5)
    
    count = 0
    myMQTTClient.connect()
    myMQTTClient.publish('mytopic',count,0)
    count = count + 1

button = Button(master, text='Click to publish', command=publish)
button.pack()
mainloop()


subscribe.py
from AWSIoTPythonSDK.MQTTLib import AWSIoTMQTTClient
import time

myMQTTClient = AWSIoTMQTTClient('Client1')
myMQTTClient.configureEndpoint('host',8883)
myMQTTClient.configureCredentials(rootCAPath, privateKeyPath, certificatePath)  
myMQTTClient.configureAutoReconnectBackoffTime(1, 32, 20)
myMQTTClient.configureOfflinePublishQueueing(-1)
myMQTTClient.configureDrainingFrequency(2)
myMQTTClient.configureConnectDisconnectTimeout(10)
myMQTTClient.configureMQTTOperationTimeout(5)


def customCallback(client, userdata, message):
    print('Received message:', message.payload)
    print('topic: ', message.topic)

myMQTTClient.connect()
myMQTTClient.subscribe('mytopic',1,customCallback)

while True:
    time.sleep(0.1)

**Comment by WALLTECH** [comment]
Hello,
I have an asynchronous pub/sub script that hlibrariesas worked for months without issues, and now fails on all wifis with this publish timeout error after around an hour of publishing every minute. It is fixed with a reboot, and any attempt to disconnect and reconnect in code throws a temporary error in name resolution. I can't share too much about the code, but I've tried many many things including starting again from a fresh OS and reinstalling, and the strangest part is it works for a while then fails every time. What should I do?

**Comment by fm-tomh** [comment]
@WALLTECH How did you go about solving this? I am experiencing the same problem, runs perfectly during testing but once its out there its struggle time. Probably something to do with the mobile network delay but no amount of playing around with AutoReconnect seems to help.

**Comment by flycast** [comment]
I wish we could be more help. We abandoned the project.

**Comment by cnrmck** [comment]
Any response on this?

**Comment by bretambrose** [comment]
This issue has been closed for quite a while and there's a lot of tangentional replies after-the-fact, so it's not easy to determine what is meant by "this".  If you're having a problem with the latest version of the SDK, please create a new issue, supplying as much detail and sample code as you can.

**Comment by cnrmck** [comment]
@bretambrose Fair. I'll open a new issue and link it here.
